Extend pending shares list in frontend to include remote shares

PVince81 opened the pull request:
And adjust the accept/decline actions to use the right endpoint for remote shares.
Note: the backend API was already available.

Fixes https://github.com/nextcloud/server/issues/23769

Also makes the following scenario work:

1. On instance A share a folder "test" with a group "group1" on instance B
2. On instance B, add a new user "new-user" to the group "group1"
3. Login as "new-user"
4. Go to "Pending shares" and accept the received remote share

In this scenario no notification is generated for the newly added user, and I don't think we want to.
This PR still makes the pending remote share visible in the correct place.

## Tests
### Single user remote share
- [x] list
- [x] accept
- [x] reject
- re-accept => not possible, once rejected it's gone, by design
- re-reject  => not possible, once rejected it's gone, by design

### Group remote share

- [x] list
- [x] accept
- [x] reject
- [x] add user to group => pending share appears for that user
- [x] re-accept
- [x] re-reject

PVince81 committed:
Extend pending shares list to include remote shares

And adjust the accept/decline actions to use the right endpoint for
remote shares.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
- [ ] bug: adjust endpoint target for rejecting share: needs extra "/pending" suffix

PVince81 commented:
- [ ] bug: some mess happening after accepting a group share, it seems the share entries have duplicated in the DB but now for the current user

I wonder if that existing backend part is really reliable or whether it contains yet to be discovered bugs :thinking:

PVince81 commented:
it looks like "oc_share_external" for groups works in a similar fashion like "oc_share": for a group share there's a main entry and sub-entries.

This is how it looks like after rejecting the group share "new-group-share-reject":
```
+----+--------+------------+--------------------------------+-----------+-----------------+----------+-------------------------+-------+-------------+-----------------------------------------------------+----------------------------------+----------+
| id | parent | share_type | remote                         | remote_id | share_token     | password | name                    | owner | user        | mountpoint                                          | mountpoint_hash                  | accepted |
+----+--------+------------+--------------------------------+-----------+-----------------+----------+-------------------------+-------+-------------+-----------------------------------------------------+----------------------------------+----------+
|  6 |     -1 |          1 | https://vvortex-release.local/ | 32        | 5HRRJLdWNSZzSmE |          | /new-group-share-accept | admin | named-users | {{TemporaryMountPointName#/new-group-share-accept}} | 3bea95b0013a55aa8d1b0fc5bd9198c9 |        0 |
|  7 |     -1 |          1 | https://vvortex-release.local/ | 33        | ynZpKQYrw5jcN9O |          | /new-group-share-reject | admin | named-users | {{TemporaryMountPointName#/new-group-share-reject}} | 3ef4d6505bba0fc9dd07a6877f387731 |        0 |
|  8 |      7 |          1 | https://vvortex-release.local/ | 33        | ynZpKQYrw5jcN9O |          | /new-group-share-reject | admin | evert       | {{TemporaryMountPointName#/new-group-share-reject}} | 3ef4d6505bba0fc9dd07a6877f387731 |        0 |
+----+--------+------------+--------------------------------+-----------+-----------------+----------+-------------------------+-------+-------------+-----------------------------------------------------+----------------------------------+----------+
```

The trouble is that all these entries are returned by the "/pending" endpoint as is, so I now see the "new-group-share-reject" entry twice in the UI. But the accept/reject actions only operate on the share id of the group share entry, the other ones return 404.

It seems that this listing endpoint is missing the logic that transforms it to the point of view of the current user. This logic should be in the backend like for regular shares. :-/

PVince81 commented:
yup, confirmed, the complex aggregation logic is missing here: https://github.com/nextcloud/server/blob/enh/23769/accept-decline-fed-sharing-ui/apps/files_sharing/lib/External/Manager.php#L644

juliushaertl commented on apps/files_sharing/js/sharedfilelist.js:
```diff
@@ -263,6 +276,7 @@
 			},
 
 			reloadCallback: function(shares, remoteShares, additionalShares) {
+				console.log('## reloadCallback', shares, remoteShares, additionalShares)
```
Leftover?

PVince81 committed:
Fix received federated group shares

Fix pending shares endpoint to consider user-specific sub-entries
for group shares whenever a share was accepted or declined.

Added unit test for adding remote group shares.

Fixed "removeUserShares" to not send a remote request as we never send
remote requests for group shares.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I've fixed the backend now and extended the unit tests, it was a bit more complicated than expected.

I've used the old style queries to save some time. At some point we should move to the query builder.

Up next: retesting through the frontend and checking for bugs

PVince81 commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -629,23 +646,39 @@ private function getShares($accepted) {
+				return !in_array($share['id'], $toRemove, true);
+			});
+
+			if (!is_null($accepted)) {
```
note: I had to move this out of the SQL query because when fetching the group share entry is never "accepted" but the sub-entry might be, and we need to have the non-accepted group share present to delete it from the results with the above filter

PVince81 commented:
fixed postgresql tests by adding more casting for the returned DB values

PVince81 commented:
- [ ] BUG: can't re-accept after rejecting group share once: this is because the share entry returned by the API is actually the sub-share, and this one can't be re-inserted...

PVince81 commented:
- [ ] make sure we only fetch a single share when processing the mount

PVince81 commented:
- [ ] remove "reject" action for pending group share as it cannot be discarded

PVince81 commented:
- [ ] prevent sidebar to open/load for remote shares as we have no data ? (click the avatar of owner)

PVince81 committed:
Remove "Reject share" for pending remote shares

In the list of pending shares, the option for rejecting the share has
been removed.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
- [ ] bug: only remove the "reject" action for received **group shares** not user shares

PVince81 committed:
Add logging to external shares manager

Instead of just returning false, also log the exception to make
debugging database issues easier.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Fix re-accepting or re-rejecting remote group shares

When accepting a group share, a sub-share entry is created which also
has a different id.

When accepting or rejecting the sub-share, simply update the "accepted"
flag instead of trying to re-insert the entry.

Adjust getShare to also properly validate group share membership
when called on a sub-share id.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I've fixed the mess in the backend and now it is possible to work with the sub-share entry when re-accepting or rejecting.

- [ ] check what happens when operating on the group share id whenever a sub-share id already exists...

PVince81 commented:
- [ ] backport/split the code that adds missing remote group share handling for status icon ?
- [ ] split the code that enhances fileactions ?

PVince81 commented:
- [ ] icon should be of a share, and maybe of a group (people + folder) for a received group share

because the user needs to understand why they can't get rid of the group share when opting out while they can for the remote user share

PVince81 committed:
Pending remote group share fixes

Only remove reject share for remote group shares
Also fix share indicator to appear for remote group shares as well.
Fix pending remote share icon to be the one of a share.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Disable default actions in pending file list

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I've fixed the issues with the file actions: the share list now has no more sidebar and the default actions are disabled, including the goto plugin. This way the user doesn't end up with a 404.

PVince81 commented:
I just re-ran `make build-js-production` and there are no local changes, and yet Node complains about core.js updates... I didn't even touch anything related. https://github.com/nextcloud/server/pull/27751/checks?check_run_id=2989100482

Maybe because I'm running node v14, I'll try again with v12...

PVince81 committed:
Fix remote group share API interactions

Accepting and declining can now be done repeatedly on both the parent
group share and sub-share with the same effects.

Added unit tests to cover these cases, and also when the same operation
is repeated.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
looks like there's no unique key that prevents two sub-shares to be created for a user with the same parent... so might need another strategy to prevent duplicate sub-shares...

PVince81 commented:
okay, problem solved now... when accessing through API we always operate on the sub-share when it exists, regardless whether the id of the parent or the sub-share was given.

PVince81 commented:
I've redone a quick test with accepting, rejecting, then reaccepting a remote group share and it worked fine.

Ready for review, now.

PVince81 commented:
would be good to backport to 22 for a future minor even though it's a bigger thing

juliusknorr reviewed (dismissed):
Tested and works :+1:

juliusknorr reviewed (commented):
Just noticed the following when trying to reject the remote group share through the notification:

```
{
  "reqId": "8fn2qjSY2jdHp9LuDMjO",
  "level": 3,
  "time": "2021-07-12T11:41:30+00:00",
  "remoteAddr": "192.168.21.6",
  "user": "admin",
  "app": "no app in context",
  "method": "DELETE",
  "url": "/ocs/v2.php/apps/files_sharing/api/v1/remote_shares/pending/4",
  "message": "An exception occurred while executing a query: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'admin-964931522bcd01381cddc4984b4e604b' for key 'sh_external_mp'",
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64; rv:89.0) Gecko/20100101 Firefox/89.0",
  "version": "23.0.0.0",
  "exception": {
    "Exception": "Doctrine\\DBAL\\Exception\\UniqueConstraintViolationException",
    "Message": "An exception occurred while executing a query: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'admin-964931522bcd01381cddc4984b4e604b' for key 'sh_external_mp'",
    "Code": 1062,
    "Trace": [
      {
        "file": "/var/www/html/3rdparty/doctrine/dbal/src/Connection.php",
        "line": 1728,
        "function": "convert",
        "class": "Doctrine\\DBAL\\Driver\\API\\MySQL\\ExceptionConverter",
        "type": "->",
        "args": [
          {
            "xdebug_message": "<tr><th align='left' bgcolor='#f57900' colspan=\"5\"><span style='background-color: #cc0000; color: #fce94f; font-size: x-large;'>( ! )</span> Doctrine\\DBAL\\Driver\\PDO\\Exception: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'admin-964931522bcd01381cddc4984b4e604b' for key 'sh_external_mp' in /var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Exception.php on line <i>26</i></th></tr>\n<tr><th align='left' bgcolor='#e9b96e' colspan='5'>Call Stack</th></tr>\n<tr><th align='center' bgcolor='#eeeeec'>#</th><th align='left' bgcolor='#eeeeec'>Time</th><th align='left' bgcolor='#eeeeec'>Memory</th><th align='left' bgcolor='#eeeeec'>Function</th><th align='left' bgcolor='#eeeeec'>Location</th></tr>\n<tr><td bgcolor='#eeeeec' align='center'>1</td><td bgcolor='#eeeeec' align='center'>0.0002</td><td bgcolor='#eeeeec' align='right'>405984</td><td bgcolor='#eeeeec'>{main}(  )</td><td title='/var/www/html/ocs/v2.php' bgcolor='#eeeeec'>.../v2.php<b>:</b>0</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>2</td><td bgcolor='#eeeeec' align='center'>0.0002</td><td bgcolor='#eeeeec' align='right'>406976</td><td bgcolor='#eeeeec'>require_once( <font color='#00bb00'>'/var/www/html/ocs/v1.php'</font> )</td><td title='/var/www/html/ocs/v2.php' bgcolor='#eeeeec'>.../v2.php<b>:</b>23</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>3</td><td bgcolor='#eeeeec' align='center'>0.1361</td><td bgcolor='#eeeeec' align='right'>5948240</td><td bgcolor='#eeeeec'>OC\\Route\\CachingRouter->match(  )</td><td title='/var/www/html/ocs/v1.php' bgcolor='#eeeeec'>.../v1.php<b>:</b>62</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>4</td><td bgcolor='#eeeeec' align='center'>0.1435</td><td bgcolor='#eeeeec' align='right'>6303288</td><td bgcolor='#eeeeec'>OC\\AppFramework\\App::main(  )</td><td title='/var/www/html/lib/private/Route/Router.php' bgcolor='#eeeeec'>.../Router.php<b>:</b>301</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>5</td><td bgcolor='#eeeeec' align='center'>0.2126</td><td bgcolor='#eeeeec' align='right'>8786408</td><td bgcolor='#eeeeec'>OC\\AppFramework\\Http\\Dispatcher->dispatch(  )</td><td title='/var/www/html/lib/private/AppFramework/App.php' bgcolor='#eeeeec'>.../App.php<b>:</b>157</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>6</td><td bgcolor='#eeeeec' align='center'>0.2164</td><td bgcolor='#eeeeec' align='right'>8840616</td><td bgcolor='#eeeeec'>OC\\AppFramework\\Http\\Dispatcher->executeController(  )</td><td title='/var/www/html/lib/private/AppFramework/Http/Dispatcher.php' bgcolor='#eeeeec'>.../Dispatcher.php<b>:</b>126</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>7</td><td bgcolor='#eeeeec' align='center'>0.2165</td><td bgcolor='#eeeeec' align='right'>8841408</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\Controller\\RemoteController->declineShare(  )</td><td title='/var/www/html/lib/private/AppFramework/Http/Dispatcher.php' bgcolor='#eeeeec'>.../Dispatcher.php<b>:</b>217</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>8</td><td bgcolor='#eeeeec' align='center'>0.2165</td><td bgcolor='#eeeeec' align='right'>8841408</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\External\\Manager->declineShare(  )</td><td title='/var/www/html/apps/files_sharing/lib/Controller/RemoteController.php' bgcolor='#eeeeec'>.../RemoteController.php<b>:</b>102</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>9</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8869248</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\External\\Manager->writeShareToDb(  )</td><td title='/var/www/html/apps/files_sharing/lib/External/Manager.php' bgcolor='#eeeeec'>.../Manager.php<b>:</b>401</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>10</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8871648</td><td bgcolor='#eeeeec'>OC\\DB\\PreparedStatement->execute(  )</td><td title='/var/www/html/apps/files_sharing/lib/External/Manager.php' bgcolor='#eeeeec'>.../Manager.php<b>:</b>221</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>11</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8871704</td><td bgcolor='#eeeeec'>Doctrine\\DBAL\\Statement->execute(  )</td><td title='/var/www/html/lib/private/DB/PreparedStatement.php' bgcolor='#eeeeec'>.../PreparedStatement.php<b>:</b>87</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>12</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8879696</td><td bgcolor='#eeeeec'>Doctrine\\DBAL\\Driver\\PDO\\Statement->execute(  )</td><td title='/var/www/html/3rdparty/doctrine/dbal/src/Statement.php' bgcolor='#eeeeec'>.../Statement.php<b>:</b>175</td></tr>\n",
            "__class__": "Doctrine\\DBAL\\Driver\\PDO\\Exception"
          },
          {
            "__class__": "Doctrine\\DBAL\\Query"
          }
        ]
      },
      {
        "file": "/var/www/html/3rdparty/doctrine/dbal/src/Connection.php",
        "line": 1667,
        "function": "handleDriverException",
        "class": "Doctrine\\DBAL\\Connection",
        "type": "->",
        "args": [
          {
            "xdebug_message": "<tr><th align='left' bgcolor='#f57900' colspan=\"5\"><span style='background-color: #cc0000; color: #fce94f; font-size: x-large;'>( ! )</span> Doctrine\\DBAL\\Driver\\PDO\\Exception: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'admin-964931522bcd01381cddc4984b4e604b' for key 'sh_external_mp' in /var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Exception.php on line <i>26</i></th></tr>\n<tr><th align='left' bgcolor='#e9b96e' colspan='5'>Call Stack</th></tr>\n<tr><th align='center' bgcolor='#eeeeec'>#</th><th align='left' bgcolor='#eeeeec'>Time</th><th align='left' bgcolor='#eeeeec'>Memory</th><th align='left' bgcolor='#eeeeec'>Function</th><th align='left' bgcolor='#eeeeec'>Location</th></tr>\n<tr><td bgcolor='#eeeeec' align='center'>1</td><td bgcolor='#eeeeec' align='center'>0.0002</td><td bgcolor='#eeeeec' align='right'>405984</td><td bgcolor='#eeeeec'>{main}(  )</td><td title='/var/www/html/ocs/v2.php' bgcolor='#eeeeec'>.../v2.php<b>:</b>0</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>2</td><td bgcolor='#eeeeec' align='center'>0.0002</td><td bgcolor='#eeeeec' align='right'>406976</td><td bgcolor='#eeeeec'>require_once( <font color='#00bb00'>'/var/www/html/ocs/v1.php'</font> )</td><td title='/var/www/html/ocs/v2.php' bgcolor='#eeeeec'>.../v2.php<b>:</b>23</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>3</td><td bgcolor='#eeeeec' align='center'>0.1361</td><td bgcolor='#eeeeec' align='right'>5948240</td><td bgcolor='#eeeeec'>OC\\Route\\CachingRouter->match(  )</td><td title='/var/www/html/ocs/v1.php' bgcolor='#eeeeec'>.../v1.php<b>:</b>62</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>4</td><td bgcolor='#eeeeec' align='center'>0.1435</td><td bgcolor='#eeeeec' align='right'>6303288</td><td bgcolor='#eeeeec'>OC\\AppFramework\\App::main(  )</td><td title='/var/www/html/lib/private/Route/Router.php' bgcolor='#eeeeec'>.../Router.php<b>:</b>301</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>5</td><td bgcolor='#eeeeec' align='center'>0.2126</td><td bgcolor='#eeeeec' align='right'>8786408</td><td bgcolor='#eeeeec'>OC\\AppFramework\\Http\\Dispatcher->dispatch(  )</td><td title='/var/www/html/lib/private/AppFramework/App.php' bgcolor='#eeeeec'>.../App.php<b>:</b>157</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>6</td><td bgcolor='#eeeeec' align='center'>0.2164</td><td bgcolor='#eeeeec' align='right'>8840616</td><td bgcolor='#eeeeec'>OC\\AppFramework\\Http\\Dispatcher->executeController(  )</td><td title='/var/www/html/lib/private/AppFramework/Http/Dispatcher.php' bgcolor='#eeeeec'>.../Dispatcher.php<b>:</b>126</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>7</td><td bgcolor='#eeeeec' align='center'>0.2165</td><td bgcolor='#eeeeec' align='right'>8841408</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\Controller\\RemoteController->declineShare(  )</td><td title='/var/www/html/lib/private/AppFramework/Http/Dispatcher.php' bgcolor='#eeeeec'>.../Dispatcher.php<b>:</b>217</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>8</td><td bgcolor='#eeeeec' align='center'>0.2165</td><td bgcolor='#eeeeec' align='right'>8841408</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\External\\Manager->declineShare(  )</td><td title='/var/www/html/apps/files_sharing/lib/Controller/RemoteController.php' bgcolor='#eeeeec'>.../RemoteController.php<b>:</b>102</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>9</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8869248</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\External\\Manager->writeShareToDb(  )</td><td title='/var/www/html/apps/files_sharing/lib/External/Manager.php' bgcolor='#eeeeec'>.../Manager.php<b>:</b>401</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>10</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8871648</td><td bgcolor='#eeeeec'>OC\\DB\\PreparedStatement->execute(  )</td><td title='/var/www/html/apps/files_sharing/lib/External/Manager.php' bgcolor='#eeeeec'>.../Manager.php<b>:</b>221</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>11</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8871704</td><td bgcolor='#eeeeec'>Doctrine\\DBAL\\Statement->execute(  )</td><td title='/var/www/html/lib/private/DB/PreparedStatement.php' bgcolor='#eeeeec'>.../PreparedStatement.php<b>:</b>87</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>12</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8879696</td><td bgcolor='#eeeeec'>Doctrine\\DBAL\\Driver\\PDO\\Statement->execute(  )</td><td title='/var/www/html/3rdparty/doctrine/dbal/src/Statement.php' bgcolor='#eeeeec'>.../Statement.php<b>:</b>175</td></tr>\n",
            "__class__": "Doctrine\\DBAL\\Driver\\PDO\\Exception"
          },
          {
            "__class__": "Doctrine\\DBAL\\Query"
          }
        ]
      },
      {
        "file": "/var/www/html/3rdparty/doctrine/dbal/src/Statement.php",
        "line": 179,
        "function": "convertExceptionDuringQuery",
        "class": "Doctrine\\DBAL\\Connection",
        "type": "->",
        "args": [
          {
            "xdebug_message": "<tr><th align='left' bgcolor='#f57900' colspan=\"5\"><span style='background-color: #cc0000; color: #fce94f; font-size: x-large;'>( ! )</span> Doctrine\\DBAL\\Driver\\PDO\\Exception: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'admin-964931522bcd01381cddc4984b4e604b' for key 'sh_external_mp' in /var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Exception.php on line <i>26</i></th></tr>\n<tr><th align='left' bgcolor='#e9b96e' colspan='5'>Call Stack</th></tr>\n<tr><th align='center' bgcolor='#eeeeec'>#</th><th align='left' bgcolor='#eeeeec'>Time</th><th align='left' bgcolor='#eeeeec'>Memory</th><th align='left' bgcolor='#eeeeec'>Function</th><th align='left' bgcolor='#eeeeec'>Location</th></tr>\n<tr><td bgcolor='#eeeeec' align='center'>1</td><td bgcolor='#eeeeec' align='center'>0.0002</td><td bgcolor='#eeeeec' align='right'>405984</td><td bgcolor='#eeeeec'>{main}(  )</td><td title='/var/www/html/ocs/v2.php' bgcolor='#eeeeec'>.../v2.php<b>:</b>0</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>2</td><td bgcolor='#eeeeec' align='center'>0.0002</td><td bgcolor='#eeeeec' align='right'>406976</td><td bgcolor='#eeeeec'>require_once( <font color='#00bb00'>'/var/www/html/ocs/v1.php'</font> )</td><td title='/var/www/html/ocs/v2.php' bgcolor='#eeeeec'>.../v2.php<b>:</b>23</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>3</td><td bgcolor='#eeeeec' align='center'>0.1361</td><td bgcolor='#eeeeec' align='right'>5948240</td><td bgcolor='#eeeeec'>OC\\Route\\CachingRouter->match(  )</td><td title='/var/www/html/ocs/v1.php' bgcolor='#eeeeec'>.../v1.php<b>:</b>62</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>4</td><td bgcolor='#eeeeec' align='center'>0.1435</td><td bgcolor='#eeeeec' align='right'>6303288</td><td bgcolor='#eeeeec'>OC\\AppFramework\\App::main(  )</td><td title='/var/www/html/lib/private/Route/Router.php' bgcolor='#eeeeec'>.../Router.php<b>:</b>301</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>5</td><td bgcolor='#eeeeec' align='center'>0.2126</td><td bgcolor='#eeeeec' align='right'>8786408</td><td bgcolor='#eeeeec'>OC\\AppFramework\\Http\\Dispatcher->dispatch(  )</td><td title='/var/www/html/lib/private/AppFramework/App.php' bgcolor='#eeeeec'>.../App.php<b>:</b>157</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>6</td><td bgcolor='#eeeeec' align='center'>0.2164</td><td bgcolor='#eeeeec' align='right'>8840616</td><td bgcolor='#eeeeec'>OC\\AppFramework\\Http\\Dispatcher->executeController(  )</td><td title='/var/www/html/lib/private/AppFramework/Http/Dispatcher.php' bgcolor='#eeeeec'>.../Dispatcher.php<b>:</b>126</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>7</td><td bgcolor='#eeeeec' align='center'>0.2165</td><td bgcolor='#eeeeec' align='right'>8841408</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\Controller\\RemoteController->declineShare(  )</td><td title='/var/www/html/lib/private/AppFramework/Http/Dispatcher.php' bgcolor='#eeeeec'>.../Dispatcher.php<b>:</b>217</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>8</td><td bgcolor='#eeeeec' align='center'>0.2165</td><td bgcolor='#eeeeec' align='right'>8841408</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\External\\Manager->declineShare(  )</td><td title='/var/www/html/apps/files_sharing/lib/Controller/RemoteController.php' bgcolor='#eeeeec'>.../RemoteController.php<b>:</b>102</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>9</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8869248</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\External\\Manager->writeShareToDb(  )</td><td title='/var/www/html/apps/files_sharing/lib/External/Manager.php' bgcolor='#eeeeec'>.../Manager.php<b>:</b>401</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>10</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8871648</td><td bgcolor='#eeeeec'>OC\\DB\\PreparedStatement->execute(  )</td><td title='/var/www/html/apps/files_sharing/lib/External/Manager.php' bgcolor='#eeeeec'>.../Manager.php<b>:</b>221</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>11</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8871704</td><td bgcolor='#eeeeec'>Doctrine\\DBAL\\Statement->execute(  )</td><td title='/var/www/html/lib/private/DB/PreparedStatement.php' bgcolor='#eeeeec'>.../PreparedStatement.php<b>:</b>87</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>12</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8879696</td><td bgcolor='#eeeeec'>Doctrine\\DBAL\\Driver\\PDO\\Statement->execute(  )</td><td title='/var/www/html/3rdparty/doctrine/dbal/src/Statement.php' bgcolor='#eeeeec'>.../Statement.php<b>:</b>175</td></tr>\n",
            "__class__": "Doctrine\\DBAL\\Driver\\PDO\\Exception"
          },
          "\n\t\t\t\tINSERT INTO `oc_share_external`\n\t\t\t\t\t(`remote`, `share_token`, `password`, `name`, `owner`, `user`, `mountpoint`, `mountpoint_hash`, `accepted`, `remote_id`, `parent`, `share_type`)\n\t\t\t\tVALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)\n\t\t\t",
          [
            "https://nextcloud.local.dev.bitgrid.net/",
            "Iw6MQuHIzsUFLzF",
            "",
            "/dfgdg",
            "user1",
            "And 7 more entries, set log level to debug to see all entries"
          ],
          []
        ]
      },
      {
        "file": "/var/www/html/lib/private/DB/PreparedStatement.php",
        "line": 87,
        "function": "execute",
        "class": "Doctrine\\DBAL\\Statement",
        "type": "->",
        "args": [
          [
            "https://nextcloud.local.dev.bitgrid.net/",
            "Iw6MQuHIzsUFLzF",
            "",
            "/dfgdg",
            "user1",
            "And 7 more entries, set log level to debug to see all entries"
          ]
        ]
      },
      {
        "file": "/var/www/html/apps/files_sharing/lib/External/Manager.php",
        "line": 221,
        "function": "execute",
        "class": "OC\\DB\\PreparedStatement",
        "type": "->",
        "args": [
          [
            "https://nextcloud.local.dev.bitgrid.net/",
            "Iw6MQuHIzsUFLzF",
            "",
            "/dfgdg",
            "user1",
            "And 7 more entries, set log level to debug to see all entries"
          ]
        ]
      },
      {
        "file": "/var/www/html/apps/files_sharing/lib/External/Manager.php",
        "line": 401,
        "function": "writeShareToDb",
        "class": "OCA\\Files_Sharing\\External\\Manager",
        "type": "->",
        "args": [
          "https://nextcloud.local.dev.bitgrid.net/",
          "Iw6MQuHIzsUFLzF",
          "",
          "/dfgdg",
          "user1",
          "admin",
          "{{TemporaryMountPointName#/dfgdg}}",
          "964931522bcd01381cddc4984b4e604b",
          0,
          "56",
          4,
          "1"
        ]
      },
      {
        "file": "/var/www/html/apps/files_sharing/lib/Controller/RemoteController.php",
        "line": 102,
        "function": "declineShare",
        "class": "OCA\\Files_Sharing\\External\\Manager",
        "type": "->",
        "args": [
          4
        ]
      },
      {
        "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
        "line": 217,
        "function": "declineShare",
        "class": "OCA\\Files_Sharing\\Controller\\RemoteController",
        "type": "->",
        "args": [
          4
        ]
      },
      {
        "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
        "line": 126,
        "function": "executeController",
        "class": "OC\\AppFramework\\Http\\Dispatcher",
        "type": "->",
        "args": [
          {
            "__class__": "OCA\\Files_Sharing\\Controller\\RemoteController"
          },
          "declineShare"
        ]
      },
      {
        "file": "/var/www/html/lib/private/AppFramework/App.php",
        "line": 157,
        "function": "dispatch",
        "class": "OC\\AppFramework\\Http\\Dispatcher",
        "type": "->",
        "args": [
          {
            "__class__": "OCA\\Files_Sharing\\Controller\\RemoteController"
          },
          "declineShare"
        ]
      },
      {
        "file": "/var/www/html/lib/private/Route/Router.php",
        "line": 301,
        "function": "main",
        "class": "OC\\AppFramework\\App",
        "type": "::",
        "args": [
          "OCA\\Files_Sharing\\Controller\\RemoteController",
          "declineShare",
          {
            "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
          },
          {
            "id": "4",
            "_route": "ocs.files_sharing.Remote.declineShare"
          }
        ]
      },
      {
        "file": "/var/www/html/ocs/v1.php",
        "line": 62,
        "function": "match",
        "class": "OC\\Route\\Router",
        "type": "->",
        "args": [
          "/ocsapp/apps/files_sharing/api/v1/remote_shares/pending/4"
        ]
      },
      {
        "file": "/var/www/html/ocs/v2.php",
        "line": 23,
        "args": [
          "/var/www/html/ocs/v1.php"
        ],
        "function": "require_once"
      }
    ],
    "File": "/var/www/html/3rdparty/doctrine/dbal/src/Driver/API/MySQL/ExceptionConverter.php",
    "Line": 60,
    "Previous": {
      "Exception": "Doctrine\\DBAL\\Driver\\PDO\\Exception",
      "Message": "SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'admin-964931522bcd01381cddc4984b4e604b' for key 'sh_external_mp'",
      "Code": 1062,
      "Trace": [
        {
          "file": "/var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Statement.php",
          "line": 84,
          "function": "new",
          "class": "Doctrine\\DBAL\\Driver\\PDO\\Exception",
          "type": "::",
          "args": [
            {
              "errorInfo": [
                "23000",
                1062,
                "Duplicate entry 'admin-964931522bcd01381cddc4984b4e604b' for key 'sh_external_mp'"
              ],
              "xdebug_message": "<tr><th align='left' bgcolor='#f57900' colspan=\"5\"><span style='background-color: #cc0000; color: #fce94f; font-size: x-large;'>( ! )</span> PDOException: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'admin-964931522bcd01381cddc4984b4e604b' for key 'sh_external_mp' in /var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Statement.php on line <i>82</i></th></tr>\n<tr><th align='left' bgcolor='#e9b96e' colspan='5'>Call Stack</th></tr>\n<tr><th align='center' bgcolor='#eeeeec'>#</th><th align='left' bgcolor='#eeeeec'>Time</th><th align='left' bgcolor='#eeeeec'>Memory</th><th align='left' bgcolor='#eeeeec'>Function</th><th align='left' bgcolor='#eeeeec'>Location</th></tr>\n<tr><td bgcolor='#eeeeec' align='center'>1</td><td bgcolor='#eeeeec' align='center'>0.0002</td><td bgcolor='#eeeeec' align='right'>405984</td><td bgcolor='#eeeeec'>{main}(  )</td><td title='/var/www/html/ocs/v2.php' bgcolor='#eeeeec'>.../v2.php<b>:</b>0</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>2</td><td bgcolor='#eeeeec' align='center'>0.0002</td><td bgcolor='#eeeeec' align='right'>406976</td><td bgcolor='#eeeeec'>require_once( <font color='#00bb00'>'/var/www/html/ocs/v1.php'</font> )</td><td title='/var/www/html/ocs/v2.php' bgcolor='#eeeeec'>.../v2.php<b>:</b>23</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>3</td><td bgcolor='#eeeeec' align='center'>0.1361</td><td bgcolor='#eeeeec' align='right'>5948240</td><td bgcolor='#eeeeec'>OC\\Route\\CachingRouter->match(  )</td><td title='/var/www/html/ocs/v1.php' bgcolor='#eeeeec'>.../v1.php<b>:</b>62</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>4</td><td bgcolor='#eeeeec' align='center'>0.1435</td><td bgcolor='#eeeeec' align='right'>6303288</td><td bgcolor='#eeeeec'>OC\\AppFramework\\App::main(  )</td><td title='/var/www/html/lib/private/Route/Router.php' bgcolor='#eeeeec'>.../Router.php<b>:</b>301</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>5</td><td bgcolor='#eeeeec' align='center'>0.2126</td><td bgcolor='#eeeeec' align='right'>8786408</td><td bgcolor='#eeeeec'>OC\\AppFramework\\Http\\Dispatcher->dispatch(  )</td><td title='/var/www/html/lib/private/AppFramework/App.php' bgcolor='#eeeeec'>.../App.php<b>:</b>157</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>6</td><td bgcolor='#eeeeec' align='center'>0.2164</td><td bgcolor='#eeeeec' align='right'>8840616</td><td bgcolor='#eeeeec'>OC\\AppFramework\\Http\\Dispatcher->executeController(  )</td><td title='/var/www/html/lib/private/AppFramework/Http/Dispatcher.php' bgcolor='#eeeeec'>.../Dispatcher.php<b>:</b>126</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>7</td><td bgcolor='#eeeeec' align='center'>0.2165</td><td bgcolor='#eeeeec' align='right'>8841408</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\Controller\\RemoteController->declineShare(  )</td><td title='/var/www/html/lib/private/AppFramework/Http/Dispatcher.php' bgcolor='#eeeeec'>.../Dispatcher.php<b>:</b>217</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>8</td><td bgcolor='#eeeeec' align='center'>0.2165</td><td bgcolor='#eeeeec' align='right'>8841408</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\External\\Manager->declineShare(  )</td><td title='/var/www/html/apps/files_sharing/lib/Controller/RemoteController.php' bgcolor='#eeeeec'>.../RemoteController.php<b>:</b>102</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>9</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8869248</td><td bgcolor='#eeeeec'>OCA\\Files_Sharing\\External\\Manager->writeShareToDb(  )</td><td title='/var/www/html/apps/files_sharing/lib/External/Manager.php' bgcolor='#eeeeec'>.../Manager.php<b>:</b>401</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>10</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8871648</td><td bgcolor='#eeeeec'>OC\\DB\\PreparedStatement->execute(  )</td><td title='/var/www/html/apps/files_sharing/lib/External/Manager.php' bgcolor='#eeeeec'>.../Manager.php<b>:</b>221</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>11</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8871704</td><td bgcolor='#eeeeec'>Doctrine\\DBAL\\Statement->execute(  )</td><td title='/var/www/html/lib/private/DB/PreparedStatement.php' bgcolor='#eeeeec'>.../PreparedStatement.php<b>:</b>87</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>12</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8879696</td><td bgcolor='#eeeeec'>Doctrine\\DBAL\\Driver\\PDO\\Statement->execute(  )</td><td title='/var/www/html/3rdparty/doctrine/dbal/src/Statement.php' bgcolor='#eeeeec'>.../Statement.php<b>:</b>175</td></tr>\n<tr><td bgcolor='#eeeeec' align='center'>13</td><td bgcolor='#eeeeec' align='center'>0.2171</td><td bgcolor='#eeeeec' align='right'>8879696</td><td bgcolor='#eeeeec'><a href='http://www.php.net/PDOStatement.execute' target='_new'>execute</a>\n(  )</td><td title='/var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Statement.php' bgcolor='#eeeeec'>.../Statement.php<b>:</b>82</td></tr>\n",
              "__class__": "PDOException"
            }
          ]
        },
        {
          "file": "/var/www/html/3rdparty/doctrine/dbal/src/Statement.php",
          "line": 175,
          "function": "execute",
          "class": "Doctrine\\DBAL\\Driver\\PDO\\Statement",
          "type": "->",
          "args": [
            [
              "https://nextcloud.local.dev.bitgrid.net/",
              "Iw6MQuHIzsUFLzF",
              "",
              "/dfgdg",
              "user1",
              "And 7 more entries, set log level to debug to see all entries"
            ]
          ]
        },
        {
          "file": "/var/www/html/lib/private/DB/PreparedStatement.php",
          "line": 87,
          "function": "execute",
          "class": "Doctrine\\DBAL\\Statement",
          "type": "->",
          "args": [
            [
              "https://nextcloud.local.dev.bitgrid.net/",
              "Iw6MQuHIzsUFLzF",
              "",
              "/dfgdg",
              "user1",
              "And 7 more entries, set log level to debug to see all entries"
            ]
          ]
        },
        {
          "file": "/var/www/html/apps/files_sharing/lib/External/Manager.php",
          "line": 221,
          "function": "execute",
          "class": "OC\\DB\\PreparedStatement",
          "type": "->",
          "args": [
            [
              "https://nextcloud.local.dev.bitgrid.net/",
              "Iw6MQuHIzsUFLzF",
              "",
              "/dfgdg",
              "user1",
              "And 7 more entries, set log level to debug to see all entries"
            ]
          ]
        },
        {
          "file": "/var/www/html/apps/files_sharing/lib/External/Manager.php",
          "line": 401,
          "function": "writeShareToDb",
          "class": "OCA\\Files_Sharing\\External\\Manager",
          "type": "->",
          "args": [
            "https://nextcloud.local.dev.bitgrid.net/",
            "Iw6MQuHIzsUFLzF",
            "",
            "/dfgdg",
            "user1",
            "admin",
            "{{TemporaryMountPointName#/dfgdg}}",
            "964931522bcd01381cddc4984b4e604b",
            0,
            "56",
            4,
            "1"
          ]
        },
        {
          "file": "/var/www/html/apps/files_sharing/lib/Controller/RemoteController.php",
          "line": 102,
          "function": "declineShare",
          "class": "OCA\\Files_Sharing\\External\\Manager",
          "type": "->",
          "args": [
            4
          ]
        },
        {
          "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
          "line": 217,
          "function": "declineShare",
          "class": "OCA\\Files_Sharing\\Controller\\RemoteController",
          "type": "->",
          "args": [
            4
          ]
        },
        {
          "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
          "line": 126,
          "function": "executeController",
          "class": "OC\\AppFramework\\Http\\Dispatcher",
          "type": "->",
          "args": [
            {
              "__class__": "OCA\\Files_Sharing\\Controller\\RemoteController"
            },
            "declineShare"
          ]
        },
        {
          "file": "/var/www/html/lib/private/AppFramework/App.php",
          "line": 157,
          "function": "dispatch",
          "class": "OC\\AppFramework\\Http\\Dispatcher",
          "type": "->",
          "args": [
            {
              "__class__": "OCA\\Files_Sharing\\Controller\\RemoteController"
            },
            "declineShare"
          ]
        },
        {
          "file": "/var/www/html/lib/private/Route/Router.php",
          "line": 301,
          "function": "main",
          "class": "OC\\AppFramework\\App",
          "type": "::",
          "args": [
            "OCA\\Files_Sharing\\Controller\\RemoteController",
            "declineShare",
            {
              "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
            },
            {
              "id": "4",
              "_route": "ocs.files_sharing.Remote.declineShare"
            }
          ]
        },
        {
          "file": "/var/www/html/ocs/v1.php",
          "line": 62,
          "function": "match",
          "class": "OC\\Route\\Router",
          "type": "->",
          "args": [
            "/ocsapp/apps/files_sharing/api/v1/remote_shares/pending/4"
          ]
        },
        {
          "file": "/var/www/html/ocs/v2.php",
          "line": 23,
          "args": [
            "/var/www/html/ocs/v1.php"
          ],
          "function": "require_once"
        }
      ],
      "File": "/var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Exception.php",
      "Line": 26,
      "Previous": {
        "Exception": "PDOException",
        "Message": "SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'admin-964931522bcd01381cddc4984b4e604b' for key 'sh_external_mp'",
        "Code": "23000",
        "Trace": [
          {
            "file": "/var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Statement.php",
            "line": 82,
            "function": "execute",
            "class": "PDOStatement",
            "type": "->",
            "args": [
              [
                "https://nextcloud.local.dev.bitgrid.net/",
                "Iw6MQuHIzsUFLzF",
                "",
                "/dfgdg",
                "user1",
                "And 7 more entries, set log level to debug to see all entries"
              ]
            ]
          },
          {
            "file": "/var/www/html/3rdparty/doctrine/dbal/src/Statement.php",
            "line": 175,
            "function": "execute",
            "class": "Doctrine\\DBAL\\Driver\\PDO\\Statement",
            "type": "->",
            "args": [
              [
                "https://nextcloud.local.dev.bitgrid.net/",
                "Iw6MQuHIzsUFLzF",
                "",
                "/dfgdg",
                "user1",
                "And 7 more entries, set log level to debug to see all entries"
              ]
            ]
          },
          {
            "file": "/var/www/html/lib/private/DB/PreparedStatement.php",
            "line": 87,
            "function": "execute",
            "class": "Doctrine\\DBAL\\Statement",
            "type": "->",
            "args": [
              [
                "https://nextcloud.local.dev.bitgrid.net/",
                "Iw6MQuHIzsUFLzF",
                "",
                "/dfgdg",
                "user1",
                "And 7 more entries, set log level to debug to see all entries"
              ]
            ]
          },
          {
            "file": "/var/www/html/apps/files_sharing/lib/External/Manager.php",
            "line": 221,
            "function": "execute",
            "class": "OC\\DB\\PreparedStatement",
            "type": "->",
            "args": [
              [
                "https://nextcloud.local.dev.bitgrid.net/",
                "Iw6MQuHIzsUFLzF",
                "",
                "/dfgdg",
                "user1",
                "And 7 more entries, set log level to debug to see all entries"
              ]
            ]
          },
          {
            "file": "/var/www/html/apps/files_sharing/lib/External/Manager.php",
            "line": 401,
            "function": "writeShareToDb",
            "class": "OCA\\Files_Sharing\\External\\Manager",
            "type": "->",
            "args": [
              "https://nextcloud.local.dev.bitgrid.net/",
              "Iw6MQuHIzsUFLzF",
              "",
              "/dfgdg",
              "user1",
              "admin",
              "{{TemporaryMountPointName#/dfgdg}}",
              "964931522bcd01381cddc4984b4e604b",
              0,
              "56",
              4,
              "1"
            ]
          },
          {
            "file": "/var/www/html/apps/files_sharing/lib/Controller/RemoteController.php",
            "line": 102,
            "function": "declineShare",
            "class": "OCA\\Files_Sharing\\External\\Manager",
            "type": "->",
            "args": [
              4
            ]
          },
          {
            "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
            "line": 217,
            "function": "declineShare",
            "class": "OCA\\Files_Sharing\\Controller\\RemoteController",
            "type": "->",
            "args": [
              4
            ]
          },
          {
            "file": "/var/www/html/lib/private/AppFramework/Http/Dispatcher.php",
            "line": 126,
            "function": "executeController",
            "class": "OC\\AppFramework\\Http\\Dispatcher",
            "type": "->",
            "args": [
              {
                "__class__": "OCA\\Files_Sharing\\Controller\\RemoteController"
              },
              "declineShare"
            ]
          },
          {
            "file": "/var/www/html/lib/private/AppFramework/App.php",
            "line": 157,
            "function": "dispatch",
            "class": "OC\\AppFramework\\Http\\Dispatcher",
            "type": "->",
            "args": [
              {
                "__class__": "OCA\\Files_Sharing\\Controller\\RemoteController"
              },
              "declineShare"
            ]
          },
          {
            "file": "/var/www/html/lib/private/Route/Router.php",
            "line": 301,
            "function": "main",
            "class": "OC\\AppFramework\\App",
            "type": "::",
            "args": [
              "OCA\\Files_Sharing\\Controller\\RemoteController",
              "declineShare",
              {
                "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
              },
              {
                "id": "4",
                "_route": "ocs.files_sharing.Remote.declineShare"
              }
            ]
          },
          {
            "file": "/var/www/html/ocs/v1.php",
            "line": 62,
            "function": "match",
            "class": "OC\\Route\\Router",
            "type": "->",
            "args": [
              "/ocsapp/apps/files_sharing/api/v1/remote_shares/pending/4"
            ]
          },
          {
            "file": "/var/www/html/ocs/v2.php",
            "line": 23,
            "args": [
              "/var/www/html/ocs/v1.php"
            ],
            "function": "require_once"
          }
        ],
        "File": "/var/www/html/3rdparty/doctrine/dbal/src/Driver/PDO/Statement.php",
        "Line": 82
      }
    },
    "CustomMessage": "--"
  }
}
```

juliushaertl commented:
Ok, seems to be only an issue if the recipient group id and user id of the accepting user matches (both admin in my case), since then the user/mountpoint unique key does conflict with the existing parent entry:

```
MariaDB [nextcloud]> select id, parent, share_type, owner, user, mountpoint, accepted from oc_share_external;
+----+--------+------------+-------+-------+------------------------------------+----------+
| id | parent | share_type | owner | user  | mountpoint                         | accepted |
+----+--------+------------+-------+-------+------------------------------------+----------+
|  4 |     -1 |          1 | user1 | admin | {{TemporaryMountPointName#/dfgdg}} |        0 |
|  7 |     -1 |          1 | admin | admin | {{TemporaryMountPointName#/Media}} |        0 |
|  9 |      7 |          1 | admin | admin | /Media                             |        1 |
| 10 |     -1 |          1 | admin | bgm   | {{TemporaryMountPointName#/Talk}}  |        0 |
| 11 |     10 |          1 | admin | user1 | {{TemporaryMountPointName#/Talk}}  |        0 |
+----+--------+------------+-------+-------+------------------------------------+----------+
```

Fine to file later for a follow up from my perspective :+1:

artonge reviewed (commented):
The frontend looks good. But I think I miss to mutch context for the PHP part. Especially with the sub and parent share part.

Also, I tried locally, but it did not work as I expected. I can see the share in "pending shares" behind the notification, but if I click on the notification's cross, the remote share disappears after a reload.

![Screenshot 2021-07-12 at 18-25-24 Pending shares - Nextcloud](https://user-images.githubusercontent.com/6653109/125323402-074f1480-e33f-11eb-8d53-728ebccc8b50.png)

kesselb commented:
The query to select the pending shares is broken if the (receiving) user is in more than one group.

PVince81 committed:
Fix external share manager with multiple user groups

Use query builder with proper matching for finding the group names.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
- [ ] test with federated shares added through public link
- [ ] test with federated shares added when no notification app is enabled

@artonge can you confirm that this is how you tested it ? I had notifications app enabled so I never see that popup dialog

PVince81 commented:
@kesselb I've pushed a fix for the multiple user group issues, thanks for spotting.

I'll rebase once I address the other reported JS issue

artonge commented:
> @artonge can you confirm that this is how you tested it ? I had notifications app enabled so I never see that popup dialog

Indeed, the notifications app helps, and it works as intended now :)

PVince81 commented on apps/files_sharing/js/sharedfilelist.js:
```diff
@@ -311,12 +337,29 @@
+							file.shareOwnerId = share.owner
+						}
+
+						if (!file.type) {
+							// pending shares usually have no type, so default to showing a directory icon
+							file.mimetype = 'dir-shared'
+						}
```
good point, it seems both should be defined: https://github.com/nextcloud/server/blob/enh/23769/accept-decline-fed-sharing-ui/apps/files_sharing/lib/Controller/RemoteController.php#L127

and "type" is more the share type than mime type

PVince81 replied:
as far as I can see here, pending shares have neither mimetype nor type:
<img width="365" alt="image" src="https://user-images.githubusercontent.com/277525/126227981-37feee42-d794-49a2-b7d6-5ea50e5aa276.png">

but now I see why your proposal makes sense, for consistency.
so let's go with that

PVince81 replied:
adjusted now

PVince81 committed:
Use mimetype instead of type when defaulting in remote shares

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
rebased and fixed the JS based on comments

next up: have a look at the test cases in https://github.com/nextcloud/server/pull/27751#issuecomment-879206836 and decide whether they are worth fixing here or separate

PVince81 commented:
- [ ] BUG: with notifications app off, the mount point is mangled:

1. disable notifications app
2. on instance B share with a group on instance A (with remote group sharing enabled on both)
3. on instance A go to the files app
4. a popup dialog appears, but don't accept the share
5. go to "pending shares"
6. accept it here
7. go to "shared with you"

Expected result: properly accepted share
Actual result: mangled name:
<img width="148" alt="image" src="https://user-images.githubusercontent.com/277525/126232178-59f6ab0f-b4b4-4550-98bc-fedc7f0ea1c8.png">

likely due to the temporary mountpoint string handling.

not sure if this is related to this PR or another issue, but I'll have a quick look still

PVince81 committed:
Fix remote group share decline+accept code path

When declining a remote group share through the dialog that appears when
notifications are off, the mount point is now correctly saved when
re-accepting.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
Seems https://github.com/nextcloud/server/pull/27751#issuecomment-882885947 was due to an uncovered scenario. I've added it to the tests and fixed it.

All should work fine now! Please re-review!

CarlSchwan commented on apps/federatedfilesharing/lib/OCM/CloudFederationProviderFiles.php:
```diff
@@ -250,7 +250,8 @@ public function shareReceived(ICloudFederationShare $share) {
 				$shareWith,
-				\OC::$server->query(IEventDispatcher::class)
+				\OC::$server->query(IEventDispatcher::class),
+				\OC::$server->getLogger()
```
```suggestion
				\OC::$server->get(LoggerInterface::class)
```

Also need to add the use statement

PVince81 replied:
cool, thanks for the hints.
I wasn't aware that it's possible to pass an exception in the LoggerInterface

CarlSchwan commented on apps/files_sharing/lib/AppInfo/Application.php:
```diff
@@ -98,7 +98,8 @@ public function __construct(array $urlParams = []) {
 				$uid,
-				$server->query(IEventDispatcher::class)
+				$server->query(IEventDispatcher::class),
+				$server->getLogger()
```
```suggestion
				$server->get(LoggerInterface::class)
```

CarlSchwan commented on apps/files_sharing/lib/Hooks.php:
```diff
@@ -43,7 +43,8 @@ public static function deleteUser($params) {
 			$params['uid'],
-			\OC::$server->query(IEventDispatcher::class)
+			\OC::$server->query(IEventDispatcher::class),
+			\OC::$server->getLogger()
```
```suggestion
			\OC::$server->get(LoggerInterface::class)
```

CarlSchwan commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -44,6 +46,7 @@
 use OCP\Http\Client\IClientService;
 use OCP\IDBConnection;
 use OCP\IGroupManager;
+use OCP\ILogger;
```
```suggestion
use use Psr\Log\LoggerInterface;
```

CarlSchwan commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -89,18 +92,24 @@ class Manager {
-								IUserManager $userManager,
-								?string $uid,
-								IEventDispatcher $eventDispatcher) {
+	/** @var ILogger */
```
```suggestion
	/** @var LoggerInterface */
```

CarlSchwan commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -89,18 +92,24 @@ class Manager {
+		IUserManager $userManager,
+		?string $uid,
+		IEventDispatcher $eventDispatcher,
+		ILogger $logger
```
```suggestion
		LoggerInterface $logger
```

CarlSchwan commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -268,21 +324,46 @@ public function acceptShare($id) {
+						$acceptShare->execute([1, $mountPoint, $hash, $subshare['id'], $this->uid]);
+						$result = true;
+					} catch (Exception $e) {
+						$this->logger->logException($e);
```
```suggestion
						$this->logger->emergency('Couldn't update share', ['exception' => $e]);
```

CarlSchwan commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -268,21 +324,46 @@ public function acceptShare($id) {
+							$share['share_type']);
+						$result = true;
+					} catch (Exception $e) {
+						$this->logger->logException($e);
```
```suggestion
						$this->logger->emergency('Couldn't create share', ['exception' => $e]);
```

CarlSchwan commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -318,23 +399,42 @@ public function declineShare($id) {
+					$this->updateAccepted((int)$subshare['id'], false);
+					$result = true;
+				} catch (Exception $e) {
+					$this->logger->logException($e);
```
```suggestion
					$this->logger->emergency('Couldn't update share', ['exception' => $e]);
```

CarlSchwan commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -318,23 +399,42 @@ public function declineShare($id) {
+						$share['share_type']);
+					$result = true;
+				} catch (Exception $e) {
+					$this->logger->logException($e);
```
```suggestion
					$this->logger->emergency('Couldn't create share', ['exception' => $e]);
```

CarlSchwan commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -519,22 +623,18 @@ public function removeShare($mountPoint): bool {
 
 			$this->removeReShares($id);
 		} catch (\Doctrine\DBAL\Exception $ex) {
+			$this->logger->logException($ex);
```
```suggestion
			$this->logger->emergency('Couldn't update share', ['exception' => $ex]);
```

CarlSchwan commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -588,7 +694,19 @@ public function removeUserShares($uid): bool {
+				$deleteResult->closeCursor();
+			}
 		} catch (\Doctrine\DBAL\Exception $ex) {
+			$this->logger->logException($ex);
```
```suggestion
				$this->logger->emergency('Couldn't get shares', ['exception' => $ex]);
```

CarlSchwan commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -629,23 +747,44 @@ private function getShares($accepted) {
+			}
+			return array_values($shares);
 		} catch (\Doctrine\DBAL\Exception $e) {
+			$this->logger->logException($e);
```
```suggestion
			$this->logger->emergency('Error when retriving shares', ['exception' => $e]);
```

PVince81 commented on apps/files_sharing/tests/External/ManagerTest.php:
```diff
@@ -112,6 +114,9 @@ protected function setUp(): void {
 			->willReturn([]);
 
+		$logger = $this->createMock(ILogger::class);
+		$logger->expects($this->never())->method('logException');
```
note to self: will also need to adjust this one

PVince81 committed:
Replace ILogger with LoggerInterface in remote share manager

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

Co-authored-by: Carl Schwan <carl@carlschwan.eu>

PVince81 commented:
one does not simply commit suggestions :sweat_smile: 
```
9) OCA\Files_Sharing\Tests\External\ManagerTest::testDeclineGroupShareAgainThroughMountPoint
ParseError: syntax error, unexpected 'use' (T_USE), expecting identifier (T_STRING) or function (T_FUNCTION) or const (T_CONST) or \\ (T_NS_SEPARATOR)
```

I'll fix it

PVince81 commented:
I've fixed/adjusted the logger interface, also in the test. The unit tests pass locally for me.

Ready for review again, thanks. :smile:

PVince81 commented on apps/files_sharing/lib/External/Manager.php:
```diff
@@ -588,7 +694,19 @@ public function removeUserShares($uid): bool {
+				// TODO: batch this with query builder
+				$query = $this->connection->prepare('
+					DELETE FROM `*PREFIX*share_external`
+					WHERE `parent` = ?
```
add "$this->uid" to properly delete it

double check this

PVince81 commented on apps/files_sharing/tests/External/ManagerTest.php:
```diff
@@ -318,19 +392,211 @@ public function testAddShare() {
+		$openShares = $this->manager->getOpenShares();
+
+		// accept through sub-share
+		$this->assertTrue($this->manager->acceptShare($groupShare['id']));
```
use $openShares for second accept

PVince81 commented on apps/files_sharing/tests/External/ManagerTest.php:
```diff
@@ -318,19 +392,211 @@ public function testAddShare() {
+		// this will return sub-entries
+		$openShares = $this->manager->getOpenShares();
+
+		// accept through sub-share
```
accept through group share

PVince81 committed:
Fix route path for pending remote shares

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
resolved conflicts and fixed route problems and missing imports for LoggerInterface

it works again.

however, while going through the code with @artonge we found some additional issues to fix

PVince81 committed:
Fix remote share deletion when deleting user

When deleting a user, we should only delete the direct remote user
shares or the remote group based subshares.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I've now fixed user deletion and also added a new method for group deletion, mostly for the sake of tests: https://github.com/nextcloud/server/pull/27751/commits/f6f2f63016e8183c7054fb28007a7d51666d8475

User deletion should now be properly limited to its share type and group subshares, and so is group deletion.

Note: I didn't add a hook yet for group deletion as it seems it will be more complicated and I didn't want to dig further in code unrelated to the task at hand.

PVince81 commented:
@artonge can you review again ?

PVince81 commented:
thanks a lot everyone! what a journey...

PVince81 commented:
/backport to stable22
